### PR TITLE
Fix redirect on content adding

### PR DESCRIPTION
**Pull request by gmelodie**
#315 introduced had an issue that made impossible for the functionality to work. Basically, the function was returning a slightly different endpoint string and we ended up redirecting to an inexistent place (hence the 404s instead of 400s when we tried to upload to  `api.estuary.tech/content/add`).

### Verification
Before (with `--disable-local-content-adding=true`):
image

After (with `--disable-local-content-adding=true`):
Screenshot from 2022-07-29 12-25-21

The behavior with `--disable-content-adding=true` (not `local`) is the same, we always reject:
image

### CODE REVIEW

**Review by en0ma — APPROVED** [review]
LGTM

**Comment by en0ma** [comment]
> I am not a big fan of us tacking on the /content/add to the shuttle endpoints. But that's a breaking change, and probably to be dealt with separately, if at all.

Me too, and I came to that conclusion as well. As is, car upload will require striping the path out. We could add a backward compatible host and users can opt-in if they want

File: handlers.go [hunk 1]
```diff
@@ -962,40 +962,40 @@ func (s *Server) handleAdd(c echo.Context, u *User) error {
...
 		return &util.HttpError{
 			Code:    http.StatusBadRequest,
 			Reason:  util.ERR_CONTENT_ADDING_DISABLED,
 			Details: "uploading content to this node is not allowed at the moment",
 		}
 	}
+	// propagate any query params
+	req, err := http.NewRequest("POST", uep[rand.Intn(len(uep))], c.Request().Body)
+	if err != nil {
+		return fmt.Errorf("failed to get preferred upload endpoints: %s", err)
```
Comment (en0ma) [inline handlers.go:982]: fix error message

File: handlers.go [hunk 2]
```diff
@@ -962,40 +962,40 @@ func (s *Server) handleAdd(c echo.Context, u *User) error {
...
+	}
+	req.Header = c.Request().Header.Clone()
+	req.URL.RawQuery = c.Request().URL.Query().Encode()
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to get preferred upload endpoints: %s", err)
```
Comment (en0ma) [inline handlers.go:989]: fix error message

File: handlers.go [hunk 3]
```diff
@@ -962,40 +962,40 @@ func (s *Server) handleAdd(c echo.Context, u *User) error {
...
+	}
+
+	c.Response().WriteHeader(resp.StatusCode)
+
+	_, err = io.Copy(c.Response().Writer, resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to get preferred upload endpoints: %s", err)
```
Comment (en0ma) [inline handlers.go:996]: fixe error message

File: handlers.go [hunk 4]
```diff
@@ -3052,7 +3052,7 @@ func (s *Server) handleGetViewer(c echo.Context, u *User) error {
 			DealDuration:          dealDuration,
 			MaxStagingWait:        maxStagingZoneLifetime,
 			FileStagingThreshold:  int64(individualDealThreshold),
-			ContentAddingDisabled: s.CM.contentAddingDisabled || u.StorageDisabled,
+			ContentAddingDisabled: s.CM.globalContentAddingDisabled || u.StorageDisabled,
```
Comment (en0ma) [inline handlers.go:3055]: this is correct for this case

File: handlers.go [hunk 5]
```diff
@@ -5316,5 +5318,5 @@ func (s *Server) getShuttleConfig(hostname string, authToken string) (interface{
 }
 
 func (s *Server) isContentAddingDisabled(u *User) bool {
-	return s.CM.contentAddingDisabled || u.StorageDisabled
+	return s.CM.globalContentAddingDisabled || u.StorageDisabled
```
Comment (en0ma) [inline handlers.go:5321]: This is not correct. for this case, we are checking if content can be added to this box - so `s.CM.contentAddingDisabled || u.StorageDisabled` is the correct way

File: main.go [hunk 6]
```diff
@@ -699,7 +699,7 @@ func main() {
 		go cm.handleShuttleMessages(cctx.Context, cfg.ShuttleMessageHandlers) // register workers/handlers to process shuttle rpc messages from a channel(queue)
 
 		// refresh pin queue for local contents
-		if !cm.contentAddingDisabled {
+		if !cm.globalContentAddingDisabled {
```
Comment (en0ma) [inline main.go:702]: this change is not correct - we do not want to bring in more content via bitswap if local adding is disable

File: replication.go [hunk 7]
```diff
@@ -302,34 +302,34 @@ func NewContentManager(db *gorm.DB, api api.Gateway, fc *filclient.FilClient, tb
...
+		pinJobs:                     make(map[uint]*pinner.PinningOperation),
+		pinMgr:                      pinmgr,
+		remoteTransferStatus:        cache,
+		shuttles:                    make(map[string]*ShuttleConnection),
+		contentSizeLimit:            util.DefaultContentSizeLimit,
+		hostname:                    cfg.Hostname,
+		inflightCids:                make(map[cid.Cid]uint),
+		FailDealOnTransferFailure:   cfg.Deal.FailOnTransferFailure,
+		isDealMakingDisabled:        cfg.Deal.Disable,
+		globalContentAddingDisabled: cfg.Content.DisableGlobalAdding,
```
Comment (en0ma) [inline replication.go:326]: thanks for making this change, I have been wanting to do it

File: handlers.go (hunk 4 above)
Comment (gmelodie) [reply handlers.go:3055]: > we are checking if the user can still upload content to any box
I think the flag `disable-content-adding` is supposed to include local, at least that's how the code seems to assume it (see https://github.com/application-research/estuary/blob/459a166316370e00aee6388e91501f5668a10d7a/handlers.go#L703-L708)

File: handlers.go (hunk 5 above)
Comment (gmelodie) [reply handlers.go:5321]: Same as above

File: handlers.go (hunk 4 above)
Comment (en0ma) [reply handlers.go:3056]: @gmelodie these are the possible outcomes

- if all 3 `s.CM.globalContentAddingDisabled`, `s.CM.localContentAddingDisabled` and `u.StorageDisabled` are `true` => `errorNoUpload`

-  if both `s.CM.globalContentAddingDisabled` and `s.CM.localContentAddingDisabled` are `true` and  `u.StorageDisabled` is `false`  => `errorNoUpload`

- if `s.CM.globalContentAddingDisabled` is `true` and `s.CM.localContentAddingDisabled` are `false` and `u.StorageDisabled` is `true` => `errorNoUpload`

- if `s.CM.globalContentAddingDisabled` is `false` and `s.CM.localContentAddingDisabled` are `true` and `u.StorageDisabled` is `true` => `errorNoUpload`

- if `s.CM.globalContentAddingDisabled` is `true` and `s.CM.localContentAddingDisabled` are `false` and `u.StorageDisabled` is `false` => `uploads to shuttle`

- if `s.CM.globalContentAddingDisabled` is `false` and `s.CM.localContentAddingDisabled` are `true` and `u.StorageDisabled` is `false` => `uploads to estuary`

Comment (en0ma) [reply handlers.go:3056]: > is supposed to include local, at least that's how the code seems to assume it see:

That was why I said **any** box - local or shuttle

File: handlers.go [hunk 8]
```diff
@@ -3052,7 +3053,7 @@ func (s *Server) handleGetViewer(c echo.Context, u *User) error {
 			DealDuration:          dealDuration,
 			MaxStagingWait:        maxStagingZoneLifetime,
 			FileStagingThreshold:  int64(individualDealThreshold),
-			ContentAddingDisabled: s.CM.contentAddingDisabled || u.StorageDisabled,
+			ContentAddingDisabled: s.CM.globalContentAddingDisabled || u.StorageDisabled,
```
Comment (en0ma) [inline handlers.go:3056]: @gmelodie you need to use `s.isContentAddingDisabled(u)` here

**Review by en0ma — APPROVED** [review]
LGTM

**Pull request merged**